zerver/tests/test_send_email.py

```diff
@@ -104,6 +113,64 @@ def test_initialize_connection(self) -> None:
         # 3 more calls to open as we try 3 times before giving up
         self.assertEqual(backend.open.call_count, 6)
 
+    def test_max_connection_lifetime(self) -> None:
+        backend = mock.MagicMock(spec=CustomSMTPEmailBackend)
+        backend.connection = mock.MagicMock(spec=SMTP)
+
+        backend.open.return_value = False
+        backend.connection.noop.return_value = [250]
+
+        # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+        with mock.patch.object(backend, "opened_at", timezone_now(), create=True), self.settings(
+            EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=0
+        ):
+            initialize_connection(backend)
+
+        self.assertEqual(backend.close.call_count, 1)
+        self.assertEqual(backend.open.call_count, 2)
+        self.assertEqual(backend.connection.noop.call_count, 0)
+
+        # Old connection is open, we will not open a new connection because max connection lifetime is None.
+        with mock.patch.object(backend, "opened_at", timezone_now(), create=True), self.settings(
+            EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=None
+        ):
+            initialize_connection(backend)
+
+        self.assertEqual(backend.close.call_count, 1)
+        self.assertEqual(backend.open.call_count, 3)
+        self.assertEqual(backend.connection.noop.call_count, 1)
+
+        # Old connection is open, we will not open a new connection because enough time has not elapsed.
+        with mock.patch.object(backend, "opened_at", timezone_now(), create=True), self.settings(
+            EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=5
+        ):
+            initialize_connection(backend)
+
+        self.assertEqual(backend.close.call_count, 1)
+        self.assertEqual(backend.open.call_count, 4)
+        self.assertEqual(backend.connection.noop.call_count, 2)
+
+        # Old connection is open, we will open a new connection because max time has elapsed.
+        with mock.patch.object(backend, "opened_at", timezone_now(), create=True), self.settings(
+            EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=5
+        ), time_machine.travel(
+            timezone_now() + timedelta(minutes=6),
+            tick=False,
+        ):
+            initialize_connection(backend)
+
+        self.assertEqual(backend.close.call_count, 2)
+        self.assertEqual(backend.open.call_count, 6)
+        self.assertEqual(backend.connection.noop.call_count, 2)
+
+        # Test opened_at is being set to timezone_now
+        with time_machine.travel(
+            timezone_now(),
+            tick=False,
+        ):
+            custom_backend = CustomSMTPEmailBackend()
+            self.assertEqual(custom_backend.opened_at, timezone_now())
+
```

Comment:
It's far more interesting to put this check inside each of your above blocks -- most notably the previous test, where you expect it to re-open the connection, and we _need_ that to reset the `opened_at`.  And using `time_machine` on all of those `initialize_connection` calls will let you know what time to expect it to be.